pkg/cloudprovider/securitygroup/securitygroup.go

```diff
@@ -101,194 +89,205 @@ Antrea internal NetworkPolicy To SecurityGroup Mapping strategy
-)
+import (
+	"fmt"
+	"sync"
 
-var (
-	ControllerPrefix             string
-	ControllerAddressGroupPrefix string
-	ControllerAppliedToPrefix    string
+	runtimev1alpha1 "antrea.io/nephe/apis/runtime/v1alpha1"
+	"antrea.io/nephe/pkg/cloudprovider/cloud"
+	"antrea.io/nephe/pkg/cloudprovider/cloudresource"
 )
 
-var ProtocolNameNumMap = map[string]int{
-	"icmp":   1,
-	"igmp":   2,
-	"tcp":    6,
-	"udp":    17,
-	"icmpv6": 58,
-}
+// CloudSecurityGroupInterface declares interface to program cloud security groups.
+type CloudSecurityGroupInterface interface {
+	// CreateSecurityGroup request to create SecurityGroup name.
+	// membershipOnly is true if the SecurityGroup is used for membership tracking, not
+	// applying ingress/egress rules.
+	// Caller expects to wait on returned channel for status
+	CreateSecurityGroup(name *cloudresource.CloudResource, membershipOnly bool) <-chan error
 
-// CloudResourceType specifies the type of cloud resource.
-type CloudResourceType string
+	// UpdateSecurityGroupRules updates SecurityGroup name's ingress/egress rules in entirety.
+	// SecurityGroup name must already been created. SecurityGroups referred to in ingressRules and
+	// egressRules must have been already created.
+	UpdateSecurityGroupRules(name *cloudresource.CloudResource, addRules, rmRules []*cloudresource.CloudRule) <-chan error
 
-var (
-	CloudResourceTypeVM  = CloudResourceType(reflect.TypeOf(runtimev1alpha1.VirtualMachine{}).Name())
-	CloudResourceTypeNIC = CloudResourceType(reflect.TypeOf(runtimev1alpha1.NetworkInterface{}).Name())
-)
+	// UpdateSecurityGroupMembers updates SecurityGroup name with members.
+	// SecurityGroup name must already have been created.
+	// For appliedSecurityGroup, UpdateSecurityGroupMembers is called only if SG has
+	// rules configured.
+	UpdateSecurityGroupMembers(name *cloudresource.CloudResource, members []*cloudresource.CloudResource, membershipOnly bool) <-chan error
+
+	// DeleteSecurityGroup deletes SecurityGroup name.
+	// SecurityGroup name must already been created, is empty.
+	DeleteSecurityGroup(name *cloudresource.CloudResource, membershipOnly bool) <-chan error
+
+	// GetSecurityGroupSyncChan returns a channel that networkPolicy controller waits on to retrieve complete SGs
+	// configured by cloud plug-in.
+	// Usage patterns:
+	// 1. Controller calls it at initialization to obtains the channel.
+	// 2. Controller waits on channel returned in 1, and expects that when channel wakes up it return the entire SGs configured.
+	// 3. Plug-in shall wake up the channel initially after sync up with the cloud; and then periodically.
+	// 4. Controller, upon receive entire SGs set, proceed to reconcile between K8s configuration and cloud configuration.
+	// This API ensures cloud plug-in stays stateless.
+	// - Correct SGs accidentally changed by customers via cloud API/console directly.
+	GetSecurityGroupSyncChan() <-chan cloudresource.SynchronizationContent
+}
+
+type SecurityGroupImpl struct{}
```

Comment:
This struct can be renamed to CloudSecurityGroupImpl as it is implementing CloudSecurityGroupInterface.
Can be done as part of subsequent cleanup.

Another type with same name(securityGroupImpl) is defined in networkpolicy.go

Reply:
Done